File: src/main/java/br/com/shoppinglist/shopping_list/shoppinglist/application/usecase/AddListItemUseCaseImpl.java
package br.com.shoppinglist.shopping_list.shoppinglist.application.usecase;

import br.com.shoppinglist.shopping_list.shoppinglist.application.dto.AddListItemCommand;
import br.com.shoppinglist.shopping_list.shoppinglist.application.port.in.AddListItemUseCase;
import br.com.shoppinglist.shopping_list.shoppinglist.application.port.out.ShoppingListRepository;
import br.com.shoppinglist.shopping_list.shoppinglist.domain.model.ShoppingListItem;
import java.util.List;
import java.util.UUID;
import org.springframework.stereotype.Service;

@Service
public class AddListItemUseCaseImpl implements AddListItemUseCase {

	private final ShoppingListRepository shoppingListRepository;

	public AddListItemUseCaseImpl(ShoppingListRepository shoppingListRepository) {
		this.shoppingListRepository = shoppingListRepository;
	}

	@Override
	public List<ShoppingListItem> execute(AddListItemCommand command) {
		if (command.description() == null || command.description().isBlank()) {
			return shoppingListRepository.findAll(command.listId());
		}

		String itemId = command.itemId() == null || command.itemId().isBlank()
			? UUID.randomUUID().toString()
			: command.itemId();

		ShoppingListItem item = new ShoppingListItem(
			itemId,
			command.description(),
			command.price(),
			command.expiry()
		);

		return shoppingListRepository.saveItem(command.listId(), item);
	}
}

File: src/shoppinglist/application/usecase/add-list-item.use-case.impl.ts
import { Injectable } from '@nestjs/common';
import { randomUUID } from 'crypto';
import { EmptyItemDescriptionException } from '../../domain/exception/shopping-list.exceptions';
import { ShoppingListItem } from '../../domain/model/shopping-list-item';
import { AddListItemCommand } from '../dto/add-list-item.command';
import { AddListItemUseCase } from '../port/in/add-list-item.use-case';
import { ShoppingListRepository } from '../port/out/shopping-list.repository';

@Injectable()
export class AddListItemUseCaseImpl extends AddListItemUseCase {
  constructor(private readonly shoppingListRepository: ShoppingListRepository) {
    super();
  }

  async execute(command: AddListItemCommand): Promise<ShoppingListItem[]> {
    const description = command.description?.trim() ?? '';
    if (!description) {
      throw new EmptyItemDescriptionException();
    }

    const itemId = command.itemId?.trim()
      ? command.itemId.trim()
      : randomUUID();
    const item: ShoppingListItem = {
      itemId,
      description,
      price: command.price ?? null,
      expiry: command.expiry ?? null,
    };

    return this.shoppingListRepository.saveItem(command.listId, item);
  }
}
